### PR TITLE
Flush stdout/err in Dockerfile-workers before replacing the current process

### LINKED ISSUE #14194: Dockerfile-workers not showing logging in early output
### Description

When running a docker image, you can watch logs of output printed(usually to /dev/stdout) by running a `docker logs container_name`. For some reason, when using the Dockerfile-workers build, you can see the output that is logged from start.py but not configure_workers_and_start.py.

### Steps to reproduce

- run docker image built from Dockerfile-workers tagged as synapse, add a `-e SYNAPSE_WORKER_TYPES=omgIneedWORKERZ` as an obvious example of an incorrect worker type
- look at output from `docker logs <container_name>`

### Homeserver

My homeserver, and Complement CI test runs

### Synapse Version

1.69.0rc2

### Installation Method

Docker (matrixdotorg/synapse)

### Platform

Docker Engine on linux with python 3.10.8

### Relevant log output

```shell
[admin@matrix2 ~]$ docker logs synapse
Generating a random secret for SYNAPSE_REGISTRATION_SHARED_SECRET
Generating a random secret for SYNAPSE_MACAROON_SECRET_KEY
Generating synapse config file /data/homeserver.yaml
Generating log config file /data/log.config
Setting ownership on /data to 991:991
2022-10-14 23:59:25,368 INFO Included extra file "/etc/supervisor/conf.d/synapse.conf" during parsing
2022-10-14 23:59:25,368 INFO Set uid to user 0 succeeded
```

### Anything else that would be useful to know?

I expected to see:
```
[admin@matrix2 ~]$ docker logs synapse
Generating base homeserver config
Generating a random secret for SYNAPSE_REGISTRATION_SHARED_SECRET
Generating a random secret for SYNAPSE_MACAROON_SECRET_KEY
Generating synapse config file /data/homeserver.yaml
Generating log config file /data/log.config
Setting ownership on /data to 991:991
omgIneedWORKERZ is an unknown worker type! It will be ignored
Starting supervisord
2022-10-15 00:09:39,165 INFO Included extra file "/etc/supervisor/conf.d/synapse.conf" during parsing
2022-10-15 00:09:39,165 INFO Set uid to user 0 succeeded
```
If this log output isn't being sent, no one will know if the system couldn't find libjemalloc.so.2, misspelled a worker_type or committed some other error while declaring variables to pass to the container during startup.

### PR DESCRIPTION
**Pull request by realtyem**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

* Fixes #14194
Additional log output there.

Some logged output from the [entrypoint](https://github.com/matrix-org/synapse/blob/develop/docker/configure_workers_and_start.py) of [Dockerfile-workers](https://github.com/matrix-org/synapse/blob/develop/docker/Dockerfile-workers) is not displayed when running a container as a background process, or detached mode, with `-d`. `-d` and detached mode is explained [here](https://docs.docker.com/engine/reference/run/#detached--d), but a quick summary is that it is used by any process or service to be run in the background and not accept input from a command line, like a standard http server. Many VPS and NAS systems(such as unRAID and fly.io) that run docker images have this option built-in to the scripting that runs docker images as 'apps'; it is not something easily changeable.

Explained somewhat [here](https://stackoverflow.com/questions/29663459/python-app-does-not-print-anything-when-running-detached-in-docker) and even more so [here](https://eklitzke.org/stdout-buffering), the culprit is because of line 'buffering' of output in Python. It appears that the buffer for stdout isn't being flushed out and it just gets 'lost'. Figuring out where it disappeared to is beyond the scope of this pull-request. The simplest solution to flushing this buffer is to include a keyword to print called `flush`. [More about print() and it's flush keyword](https://docs.python.org/3.10/library/functions.html#print). The keyword was added in Python 3.3 and therefore isn't in danger of being non-existent for any versions of Python used or tested today.

As observed inside of [start.py](https://github.com/matrix-org/synapse/blob/develop/docker/start.py#L16), the issue is mitigated by sending logged output to stderr instead of stdout. This works, but doesn't feel like the right way to handle this. Just flush the buffers. If someone would rather I switch to using stderr, leave me a comment to that effect and I'll adjust.

Other options explored:
- PYTHONUNBUFFERED=1 in the environment before starting the Python process.
- Starting the python scripts with `-u`, such as `python -u configure_workers_and_start.py`

Neither seemed viable and felt intrusive to fix one tiny oversight in the way buffered output is handled.

There are two commits in this pull request that can be reverted. The one is harmless and probably would be good to keep. The other is bad and shouldn't be used long term on Github(tons of log spam):
- A line is output when `Generating base homeserver config` that didn't have a mate when config already exists. I added that in before collecting logs while diagnosing the issue. Look for `Homeserver config already exists at:`.
- Added the environment variable `COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1` to the complement CI workflow so when tests gets ran, you won't have to wait for a flake or an error to actually see the log output.

Github CI Complement snippet output from before:
```
Generating RSA private key, 2048 bit long modulus (2 primes)
...........................+++++
...........................................+++++
e is 65537 (0x010001)
Signature ok
subject=CN = hs1
Getting CA Private Key
                DNS:hs1
2022-10-16 00:57:23,912 INFO Included extra file "/etc/supervisor/conf.d/postgres.conf" during parsing
2022-10-16 00:57:23,912 INFO Included extra file "/etc/supervisor/conf.d/synapse.conf" during parsing
2022-10-16 00:57:23,912 INFO Set uid to user 0 succeeded
2022-10-16 00:57:23,914 INFO supervisord started with pid 1
```
After:
```
Generating RSA private key, 2048 bit long modulus (2 primes)
....+++++
....................................................................+++++
e is 65537 (0x010001)
Signature ok
subject=CN = hs1
Getting CA Private Key
                DNS:hs1
Homeserver config already exists at: /data/homeserver.yaml
Starting supervisord
2022-10-16 01:19:35,454 INFO Included extra file "/etc/supervisor/conf.d/postgres.conf" during parsing
2022-10-16 01:19:35,454 INFO Included extra file "/etc/supervisor/conf.d/synapse.conf" during parsing
2022-10-16 01:19:35,454 INFO Set uid to user 0 succeeded
2022-10-16 01:19:35,456 INFO supervisord started with pid 1
```
And yes it's literally only two lines different. Apparently, the Github Complement suites generate their own configuration elsewhere or hide the log output of start.py if it's ran. This doesn't happen with the command-line ran [script](https://github.com/matrix-org/synapse/blob/develop/scripts-dev/complement.sh). That I tested by creating an obviously wrong worker type and feeding it through(See issue at top). The configuration was not created(which is the correct response) but because that part of [configure_workers_and_start.py]() never gets touched(at least in these logs), it didn't produce the information that it was an incorrect worker type. I'm assuming that is something configurable with GO or Complement and is a rabbit hole for another day.

Output in `docker logs synapse` works as expected.

Signed-off-by: Jason Little <realtyem@gmail.com>

### CODE REVIEW

**Comment by realtyem** [comment]
Spent a few minutes tracking down the fail. Looks like the Alias flake
```
2022-10-16T03:02:51.3775894Z ##[group][0;31m❌ TestRoomAlias[0;37m (36.99s)[0m
2022-10-16T03:02:51.3776170Z 2022/10/16 02:46:16 Sharing [SERVER_NAME=hs1 SYNAPSE_COMPLEMENT_USE_WORKERS=true SYNAPSE_COMPLEMENT_DATABASE=postgres] host environment variables with container
2022-10-16T03:02:51.3776434Z 2022/10/16 02:46:34 Sharing [SERVER_NAME=hs1 SYNAPSE_COMPLEMENT_USE_WORKERS=true SYNAPSE_COMPLEMENT_DATABASE=postgres] host environment variables with container
2022-10-16T03:02:51.3776600Z     apidoc_room_alias_test.go:43: Deploy times: 18.08280097s blueprints, 15.881970214s containers
2022-10-16T03:02:51.3776773Z     client.go:604: [CSAPI] POST hs1/_matrix/client/v3/createRoom => 200 OK (666.647458ms)
2022-10-16T03:02:51.3776939Z     client.go:604: [CSAPI] POST hs1/_matrix/client/v3/createRoom => 200 OK (735.356276ms)
2022-10-16T03:02:51.3777132Z     client.go:604: [CSAPI] PUT hs1/_matrix/client/v3/directory/room/#creates_alias:hs1 => 200 OK (80.10207ms)
2022-10-16T03:02:51.3777373Z     client.go:604: [CSAPI] GET hs1/_matrix/client/v3/directory/room/#creates_alias:hs1 => 200 OK (11.070455ms)
2022-10-16T03:02:51.3777568Z     client.go:604: [CSAPI] PUT hs1/_matrix/client/v3/directory/room/#room_members_list:hs1 => 200 OK (22.387208ms)
2022-10-16T03:02:51.3777758Z     client.go:604: [CSAPI] GET hs1/_matrix/client/v3/directory/room/#room_members_list:hs1 => 200 OK (7.739968ms)
2022-10-16T03:02:51.3777970Z     client.go:604: [CSAPI] GET hs1/_matrix/client/v3/rooms/!VPygoawTzSovztdqiv:hs1/aliases => 403 Forbidden (20.338016ms)
2022-10-16T03:02:51.3778134Z     client.go:604: [CSAPI] POST hs1/_matrix/client/v3/createRoom => 200 OK (788.782257ms)
2022-10-16T03:02:51.3778333Z     client.go:604: [CSAPI] GET hs1/_matrix/client/v3/rooms/!yOMojKJgXDwFkaEYcv:hs1/aliases => 200 OK (11.944651ms)
2022-10-16T03:02:51.3778517Z     client.go:604: [CSAPI] PUT hs1/_matrix/client/v3/directory/room/#lists_aliases:hs1 => 200 OK (13.035746ms)
2022-10-16T03:02:51.3778716Z     client.go:604: [CSAPI] GET hs1/_matrix/client/v3/rooms/!yOMojKJgXDwFkaEYcv:hs1/aliases => 200 OK (5.369578ms)
2022-10-16T03:02:51.3778939Z ##[endgroup]
2022-10-16T03:02:51.3779257Z ##[group][0;31m❌ TestRoomAlias/Parallel[0;37m (0s)[0m
2022-10-16T03:02:51.3779463Z ##[endgroup]
2022-10-16T03:02:51.3779840Z ##[group][0;31m❌ TestRoomAlias/Parallel/GET_/rooms/:room_id/aliases_lists_aliases[0;37m (820ms)[0m
2022-10-16T03:02:51.3780337Z     apidoc_room_alias_test.go:87: MatchResponse key 'aliases' got '[]' want '[#lists_aliases:hs1]' - http://localhost:49277/_matrix/client/v3/rooms/%21yOMojKJgXDwFkaEYcv:hs1/aliases => {"aliases":[]}
2022-10-16T03:02:51.3780445Z 2022/10/16 02:46:52 ============================================
2022-10-16T03:02:51.3780454Z 
```

Text searching the raw logs for the ❌ found the error. The logs are huge, which is why I recommend reverting that 'feature'.

File: docker/start.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 # Utility functions
 def log(txt: str) -> None:
-    print(txt, file=sys.stderr)
+    print(txt, flush=True)
```
Comment (DMRobertson) [inline docker/start.py:16]: `start.py` is also used by the non-worker Dockerfile. Is this change intentional here, given that you are only targetting Dockerfile-workers?

Note that this change also makes it so that the `error` function below writes to stdout, not stderr---which sounds wrong to me.

File: changelog.d/14195.docker [hunk 2]
```diff
@@ -0,0 +1 @@
+Fix early logging from Dockerfile-workers image.
```
Comment (DMRobertson) [inline changelog.d/14195.docker:1]:
```suggestion
Fix pre-startup logging being lost when using the `Dockerfile-workers` image.
```

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Not sure about some of this.

I think a cleaner fix might be to call `sys.stdout.flush()` and `sys.stderr.flush()` before we call `os.execle`/`os.execv`/`os.execve`?

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -465,6 +465,8 @@ jobs:
           POSTGRES=${{ (matrix.database == 'Postgres') && 1 || '' }} WORKERS=${{ (matrix.arrangement == 'workers') && 1 || '' }} COMPLEMENT_DIR=`pwd`/complement synapse/scripts-dev/complement.sh -json 2>&1 | synapse/.ci/scripts/gotestfmt
         shell: bash
         name: Run Complement Tests
+        env:
+          COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS: 1
```
[the remark this answers is not in the document]
Comment (realtyem) [reply .github/workflows/tests.yml:469]: It is.
I mention reverting it once everything is reconciled.
>There are two commits in this pull request that can be reverted. The one is harmless and probably would be good to keep. The other is bad and shouldn't be used long term on Github(tons of log spam):

>* A line is output when Generating base homeserver config that didn't have a mate when config already exists. I added that in before collecting logs while diagnosing the issue. Look for Homeserver config already exists at:.
>* Added the environment variable COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 to the complement CI workflow so when tests gets ran, you won't have to wait for a flake or an error to actually see the log output.

And will be happy to do so. The log spam is outrageous, but was needed for log proof since flakes are random and I didn't want to rely on one just get to the actual server logs

File: docker/configure_workers_and_start.py [hunk 4]
```diff
@@ -619,6 +619,8 @@ def main(args: List[str], environ: MutableMapping[str, str]) -> None:
     if not os.path.exists(config_path):
         log("Generating base homeserver config")
         generate_base_homeserver_config()
+    else:
+        log("Homeserver config already exists at: " + config_path)
```
[the remark this answers is not in the document]
Comment (realtyem) [reply docker/configure_workers_and_start.py:623]: Same as above. The commit that made it even says it can be reverted. I mention it in the original comment.
>There are two commits in this pull request that can be reverted. The one is harmless and probably would be good to keep. The other is bad and shouldn't be used long term on Github(tons of log spam):

>* A line is output when Generating base homeserver config that didn't have a mate when config already exists. I added that in before collecting logs while diagnosing the issue. Look for Homeserver config already exists at:.
>* Added the environment variable COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 to the complement CI workflow so when tests gets ran, you won't have to wait for a flake or an error to actually see the log output.

And I agree about extra logging not being here. As a curiousity, is that appropriate in an isolated pull request?

File: docker/start.py (hunk 1 above)
Comment (realtyem) [reply docker/start.py:16]: [Docker Documentation](https://docs.docker.com/config/containers/logging/).
>`STDOUT` is usually a command’s normal output, and `STDERR` is typically used to output error messages. By default, docker logs shows the command’s `STDOUT` and `STDERR`.

I was only trying to make sure it all worked the same way. The `error()` subroutine prints and exits, in both files. If you like, put it back the way it was. I did offer that above in the original comment. But it did feel wrong to me for `log()` to print to stderr by default.
>If someone would rather I switch to using stderr, leave me a comment to that effect and I'll adjust.

Perhaps a compromise of:  (from start.py)

```sh
# Utility functions
def log(txt: str) -> None:
    print(txt, flush=True)


def error(txt: str) -> NoReturn:
    print(txt, file=sys.stderr)
    sys.exit(2)
```

Duplicated across both? Maintaining the comments in configure....py(are those what are called docstrings?) of course.

**Comment by realtyem** [comment]
> Not sure about some of this.
> 
> I think a cleaner fix might be to call `sys.stdout.flush()` and `sys.stderr.flush()` before we call `os.execle`/`os.execv`/`os.execve`?

I admit to coming across those in my research about this. I did not think they would work without a companion `sys.stdout.write()` and friend. Turns out they do.

A quick set of tests yielded odd results. I won't rehash the bulk of what I tried out in combinations but the most notable was that leaving `start.py` with a `flush=True` and removing it in `configure_workers_and_start.py` but adding a `sys.stdout.flush()` right above the `os.execle()` yielded log output from `configure_workers_and_start.py`, but nothing from `start.py`. Very strange, that.

I'm pretty sure I stand by my change here. Having `flush=True` in start.py means not having to add a  `sys.stdout.flush()` call in three places. Isn't one keyword argument change better than three other lines added, just to fix an anomaly in buffer handling when that option is available?

File: .github/workflows/tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (DMRobertson) [reply .github/workflows/tests.yml:469]: Sorry, I should have read your description more carefully.

**Comment by realtyem** [comment]
If you like I can include a comment above the change about why the flush was added?

File: docker/configure_workers_and_start.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (realtyem) [reply docker/configure_workers_and_start.py]: You got my meaning exactly. Thanks.

File: .github/workflows/tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (realtyem) [reply .github/workflows/tests.yml]: I tried to be as verbose as I could manage without beating my head against a wall. I spent hours on that 😄

**Comment by DMRobertson** [comment]
> Having `flush=True` in start.py means not having to add a `sys.stdout.flush()` call in three places. Isn't one keyword argument change better than three other lines added[...]?

Yes and no. If someone in the future writes makes a `print()` call without `flush`ing, or starts doing so indirectly via Python's stdlib logging, we're going to get the same problem all over again.

> just to fix an anomaly in buffer handling when that option is available?

I don't think this is an anomaly in buffer handling; this is something that you simply are expected to do when replacing the current process with an `exec*` call. The Python [docs](https://docs.python.org/3/library/os.html#os.execl) even highlight this:

> The current process is replaced immediately. Open file objects and descriptors are not flushed, so if there may be data buffered on these open files, you should flush them using sys.stdout.flush() or [os.fsync()](https://docs.python.org/3/library/os.html#os.fsync) before calling an [exec*](https://docs.python.org/3/library/os.html#os.execl) function.

So I think explicitly flushing before exec-ing is the more robust (and arguably idiomatic) change here.

**Comment by realtyem** [comment]
> > Having `flush=True` in start.py means not having to add a `sys.stdout.flush()` call in three places. Isn't one keyword argument change better than three other lines added[...]?
> 
> Yes and no. If someone in the future writes makes a `print()` call without `flush`ing, or starts doing so indirectly via Python's stdlib logging, we're going to get the same problem all over again.

Well. I suppose that's a good point. I would argue if they've heard of those things they would look and see what other's have done before in a file they are changing. But, still a valid point.
> 
> > just to fix an anomaly in buffer handling when that option is available?
> 
> I don't think this is an anomaly in buffer handling; this is something that you simply are expected to do when replacing the current process with an `exec*` call. The Python [docs](https://docs.python.org/3/library/os.html#os.execl) even highlight this:
> 
> > The current process is replaced immediately. Open file objects and descriptors are not flushed, so if there may be data buffered on these open files, you should flush them using sys.stdout.flush() or [os.fsync()](https://docs.python.org/3/library/os.html#os.fsync) before calling an [exec*](https://docs.python.org/3/library/os.html#os.execl) function.
> 
> So I think explicitly flushing before exec-ing is the more robust (and arguably idiomatic) change here.

And this is me learning something new today. I had not come across that detail in previous research. It would have saved me a few hours, I'm sure. Very well, I'll add `sys.stdout.flush()` to start.py  and `configure_workers_and_start.py` and leave them both writing to `stdout` but without a `flush=True`. Sound good?

File: docker/start.py (hunk 1 above)
Comment (realtyem) [reply docker/start.py:16]: Nevermind, discussion continued below.

**Comment by realtyem** [comment]
Or perhaps just make both of them `print(str, file=sys.stderr)`. Somehow that feels wrong to me, but results 🤷‍♂️

**Comment by DMRobertson** [comment]
> I'll add sys.stdout.flush() to start.py and configure_workers_and_start.py and leave them both writing to stdout but without a flush=True. Sound good?

I think having `error()` write to sys.stderr would be a nice change too while we're here---though you'd need to `sys.stderr.flush()` too before an `os.exec*` call. Also beware that there are three different places where we call `os.exec*` in `start.py`.

Otherwise sounds good!

**Comment by realtyem** [comment]
> > I'll add sys.stdout.flush() to start.py and configure_workers_and_start.py and leave them both writing to stdout but without a flush=True. Sound good?
> 
> I think having `error()` write to sys.stderr would be a nice change too while we're here---though you'd need to `sys.stderr.flush()` too before an `os.exec*` call. Also beware that there are three different places where we call `os.exec*` in `start.py`.
> 
> Otherwise sounds good!

I went for(essentially):
```
def log(txt: str) -> None:
    print(txt)


def error(txt: str) -> NoReturn:
    print(txt, file=sys.stderr)
    sys.exit(2)
```
and then added:
```
sys.stdout.flush()
sys.stderr.flush()
```
above each `os.exec` variant. Testing this really quick then showed that start.py outputted no logs, as it gets called with an argument that returns without using any `os.exec()` variant.
* Inside `configure_workers_and_start.py`, [generate_base_homeserver_config()](https://github.com/matrix-org/synapse/blob/develop/docker/configure_workers_and_start.py#L322) calls out to `start.py` with a `subprocess.check_output()` using `["/usr/local/bin/python", "/start.py", "migrate_config"]`
* In start.py, it bounces from main into [this](https://github.com/matrix-org/synapse/blob/develop/docker/start.py#L210) conditional which then calls [`generate_config_from_template()](https://github.com/matrix-org/synapse/blob/develop/docker/start.py#L39)`
* Inside `generate_config_from_template()`, a set of arguments is constructed, then passed to `subprocess.check_output()` at which point it returns all the way back to `configure_workers_and_start.py` and continues onwards to finish.
* Poof! Buffers from `stdout` gone.

Would you like another set of `flush()`s above the `subprocess.check_output()` in start.py(and any more of those I find while I'm still in there)?

**Comment by DMRobertson** [comment]
> Inside configure_workers_and_start.py, [generate_base_homeserver_config()](https://github.com/matrix-org/synapse/blob/develop/docker/configure_workers_and_start.py?rgh-link-date=2022-10-17T12%3A15%3A50Z#L322) calls out to start.py with a subprocess.check_output() using ["/usr/local/bin/python", "/start.py", "migrate_config"]

[check_output](https://docs.python.org/3/library/subprocess.html?highlight=subprocess#subprocess.check_output):
- spawns an entirely new process with an entirely new stdout/stderr stream
- these streams are completely disconnected from the controlling process's stdout/stderr
- for check_output specifically, the controlling process captures the subprocess's stdout; its stderr is discarded

Try changing `check_output` to `check_call`. See https://docs.python.org/3/library/subprocess.html for more details.

**Comment by realtyem** [comment]
> > Inside configure_workers_and_start.py, [generate_base_homeserver_config()](https://github.com/matrix-org/synapse/blob/develop/docker/configure_workers_and_start.py?rgh-link-date=2022-10-17T12%3A15%3A50Z#L322) calls out to start.py with a subprocess.check_output() using ["/usr/local/bin/python", "/start.py", "migrate_config"]
> 
> [check_output](https://docs.python.org/3/library/subprocess.html?highlight=subprocess#subprocess.check_output):
> 
> * spawns an entirely new process with an entirely new stdout/stderr stream
> * these streams are completely disconnected from the controlling process's stdout/stderr
> * for check_output specifically, the controlling process captures the subprocess's stdout; its stderr is discarded
> 
> Try changing `check_output` to `check_call`. See https://docs.python.org/3/library/subprocess.html for more details.

`check_call` worked. I'll fix this up later tonight and send it in then.

**Comment by realtyem** [comment]
Any objections to a helper function? Here's what I'm looking at duplicating across both files:
```python
 Utility functions
def log(txt: str) -> None:
    """
    Log something to the stdout.

    Args:
        txt: The text to log.
    """
    print(txt)


def error(txt: str) -> NoReturn:
    """
    Log something to the stderr and exit with an error code.

    Args:
        txt: The text to log in error.
    """
    print(txt, file=sys.stderr)
    sys.exit(2)


def flush_buffers() -> None:
    """
    Flush stdout and stderr buffers
    """
    sys.stdout.flush()
    sys.stderr.flush()
```

Then can just call `flush_buffers()` at each exit point.

**Comment by realtyem** [comment]
Any objections to using `subprocess.run()` instead of `subprocess.check_call()`? I was reading about them in the link you gave, https://docs.python.org/3/library/subprocess.html, and it was said:
>Code needing to capture stdout or stderr should use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) instead:

So I read a bit on `run()`, and it's more...robust? Also the 
>Older high-level API

title above all the `subprocess.check_`* calls seems worrying. In testing this out, `run()` seems to work fine. Is there a higher level process somewhere(like maybe in Complement) that depends on using `subprocess.check_`* calls?

**Comment by DMRobertson** [comment]
These are simple bootstrapping scripts: we don't need thorough docstrings here (particularly when the functions are short and the docstrings are longer than the functions they describe). Otherwise no objections to helpers.

`check_output` etc are implemented on top of `run()`, I think; there's no robustness difference AFAIK. They're oldish, but not going to be removed any time soon (no deprecation notice; the docs even note that ` lots of existing code calls these functions`). If you particularly want to use `run` I won't stop you; crucially it's present in all Python versions we support (3.7+ at the time of writing). But I don't think it's a must-have.

**Comment by realtyem** [comment]
They were in `configure_workers_and_start.py`, just thought I would for consistency. I'll use `run()` but take out the docstrings.

File: docker/start.py [hunk 5]
```diff
@@ -131,10 +136,10 @@ def generate_config_from_template(
 
     if ownership is not None:
         log(f"Setting ownership on /data to {ownership}")
-        subprocess.check_output(["chown", "-R", ownership, "/data"])
+        subprocess.run(["chown", "-R", ownership, "/data"])
```
Comment (DMRobertson) [inline docker/start.py:139]: You need to pass `check=True` to `run()` or else we will silently ignore commands that fail.

See the docs for check_output and check_call.

Comment (realtyem) [reply docker/start.py:139]: Just to be clear, all of them? Or just those like above where it accesses outside OS commands? There are a total of 4 between both files.

Comment (DMRobertson) [reply docker/start.py:139]: All of them---we want errors to be reported loudly (as is the case with `check_output` and `check_call`).

Comment (realtyem) [reply docker/start.py:139]: I guess they all sort of do, don't they.
